src/main/java/org/springframework/integration/aws/outbound/KinesisMessageHandler.java

```diff
@@ -245,4 +295,74 @@ private PutRecordRequest buildPutRecordRequest(Message<?> message) {
 				.withData(data);
 	}
 
+	@SuppressWarnings("unchecked")
+	private AsyncHandler<PutRecordsRequest, PutRecordsResult> getPutRecordsAsyncHandler(final Message<?> message) {
+		// allowing explicit setting of the AsyncHandler for backward compatibility
+		if (this.asyncHandler != null) {
+			return (AsyncHandler<PutRecordsRequest, PutRecordsResult>) this.asyncHandler;
+		}
+		else {
+			if (getSendFailureChannel() != null || getOutputChannel() != null) {
+				return new AsyncHandler<PutRecordsRequest, PutRecordsResult>() {
+
+					@Override
+					public void onError(Exception ex) {
+						if (getSendFailureChannel() != null) {
+							KinesisMessageHandler.this.messagingTemplate.send(getSendFailureChannel(),
+									KinesisMessageHandler.this.errorMessageStrategy.buildErrorMessage(
+											new AwsSendFailureException(message,
+													(PutRecordsRequest) message.getPayload(), ex), null));
+						}
+					}
+
+					@Override
+					public void onSuccess(PutRecordsRequest request, PutRecordsResult putRecordsResult) {
+						if (getOutputChannel() != null) {
+							KinesisMessageHandler.this.messagingTemplate.send(getOutputChannel(),
+									getMessageBuilderFactory().fromMessage(message).build());
+						}
+					}
+				};
+			}
+			else {
+				return null;
+			}
+		}
+	}
+
+	@SuppressWarnings("unchecked")
+	private AsyncHandler<PutRecordRequest, PutRecordResult> getPutRecordAsyncHandler(final Message<?> message,
+																					final PutRecordRequest request) {
+		// allowing explicit setting of the AsyncHandler for backward compatibility
+		if (this.asyncHandler != null) {
+			return (AsyncHandler<PutRecordRequest, PutRecordResult>) this.asyncHandler;
+		}
+		else {
+			if (getSendFailureChannel() != null || getOutputChannel() != null) {
```

Comment:
Since we always will delegate to the user-provided `this.asyncHandler`, we don't need this check and we always will populate our own `AsyncHandler` even if its methods bodies are going to be empty.